pkg/oidc/provider/userinfo.go

```diff
@@ -0,0 +1,125 @@
+package provider
+
+import (
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"slices"
+	"strings"
+
+	"github.com/golang-jwt/jwt"
+	wrangmgmtv3 "github.com/rancher/rancher/pkg/generated/controllers/management.cattle.io/v3"
+	oidcerror "github.com/rancher/rancher/pkg/oidc/provider/error"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+)
+
+// UserInfoResponse represents the response from the userinfo endpoint.
+type UserInfoResponse struct {
+	Sub      string   `json:"sub"`
+	UserName string   `json:"username,omitempty"`
+	Groups   []string `json:"groups,omitempty"`
+}
+
+type userInfoHandler struct {
+	userCache           wrangmgmtv3.UserCache
+	userAttributeLister wrangmgmtv3.UserAttributeCache
+	jwks                signingKeyGetter
+}
+
+func newUserInfoHandler(userLister wrangmgmtv3.UserCache, userAttributeLister wrangmgmtv3.UserAttributeCache, jwks signingKeyGetter) *userInfoHandler {
+
+	return &userInfoHandler{
+		userCache:           userLister,
+		userAttributeLister: userAttributeLister,
+		jwks:                jwks,
+	}
+}
+
+// userInfoEndpoint handles the userinfo endpoint of the OIDC provider
+func (h *userInfoHandler) userInfoEndpoint(w http.ResponseWriter, r *http.Request) {
+	accessToken, err := getTokenFromHeader(r)
+	claims := jwt.MapClaims{}
+	// verify access_token signature
+	_, err = jwt.ParseWithClaims(accessToken, &claims, func(token *jwt.Token) (interface{}, error) {
+		// Ensure correct signing method
+		if _, ok := token.Method.(*jwt.SigningMethodRSA); !ok {
+			return nil, fmt.Errorf("unexpected signing method: %v", token.Header["alg"])
+		}
+		kid, ok := token.Header["kid"].(string)
+		if !ok {
+			return nil, fmt.Errorf("can't find kid")
+		}
+		pubKey, err := h.jwks.GetPublicKey(kid)
+		if err != nil {
+			return nil, err
+		}
+
+		return pubKey, nil
+	})
+	if err != nil {
+		oidcerror.WriteError(oidcerror.InvalidRequest, fmt.Sprintf("invalid access_token: %v", err), http.StatusBadRequest, w)
+		return
+	}
+
+	userId, ok := claims["sub"].(string)
+	if !ok {
+		oidcerror.WriteError(oidcerror.InvalidRequest, "invalid access_token: it doesn't have sub", http.StatusBadRequest, w)
+		return
+	}
```

Comment:
Just as a note, if the assertion fails it could also means the value was of another type (https://go.dev/play/p/b_g8YGkJhbd).

Reply:
we create the `access_token` as part of the token endpoint response, and it has to be a string. If it is not a string is not a valid token